analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantNullCheck.cs

```diff
@@ -86,21 +77,114 @@ protected override SyntaxNode GetIsOperatorCheckVariable(SyntaxNode node)
+            var left = binaryPatternNode.Left.SyntaxNode.RemoveParentheses();
+            var right = binaryPatternNode.Right.SyntaxNode.RemoveParentheses();
+
+            if (IsNotNullPattern(left) && IsAffirmativePatternMatch(right))
+            {
+                context.ReportDiagnosticWhenActive(Diagnostic.Create(SupportedDiagnostics[0], left.GetLocation()));
+            }
+            else if (IsNotNullPattern(right) && IsAffirmativePatternMatch(left))
+            {
+                context.ReportDiagnosticWhenActive(Diagnostic.Create(SupportedDiagnostics[0], right.GetLocation()));
+            }
+
+            static bool IsNotNullPattern(SyntaxNode node) =>
+                UnaryPatternSyntaxWrapper.IsInstance(node)
+                && ((UnaryPatternSyntaxWrapper)node) is var unaryPatternSyntaxWrapper
+                && unaryPatternSyntaxWrapper.IsNotNull();
+
+            // Verifies the given pattern is an affirmative pattern - constant pattern (except 'null'), Declaration pattern, Recursive pattern.
+            // The PatternSyntax appears e.g. in switch arms and is different from IsPatternSyntax.
+            static bool IsAffirmativePatternMatch(SyntaxNode node) =>
+                PatternSyntaxWrapper.IsInstance(node)
+                && ((PatternSyntaxWrapper)node) is var isPatternWrapper
+                && !isPatternWrapper.IsNot()
+                && !isPatternWrapper.IsNull();
+        }
+
+        private void CheckOrPattern(SyntaxNodeAnalysisContext context)
+        {
+            var binaryPatternNode = (BinaryPatternSyntaxWrapper)context.Node;
+            var left = binaryPatternNode.Left.SyntaxNode.RemoveParentheses();
+            var right = binaryPatternNode.Right.SyntaxNode.RemoveParentheses();
+            if (PatternSyntaxWrapper.IsInstance(left) && PatternSyntaxWrapper.IsInstance(right))
+            {
+                var leftPattern = (PatternSyntaxWrapper)left;
+                var rightPattern = (PatternSyntaxWrapper)right;
+                if (leftPattern.IsNull() && IsNegativePatternMatch(rightPattern))
+                {
+                    context.ReportDiagnosticWhenActive(Diagnostic.Create(SupportedDiagnostics[0], left.GetLocation()));
+                }
+                else if (rightPattern.IsNull() && IsNegativePatternMatch(leftPattern))
+                {
+                    context.ReportDiagnosticWhenActive(Diagnostic.Create(SupportedDiagnostics[0], right.GetLocation()));
+                }
+            }
+
+            // Verifies if the given pattern syntax is like "not X" - where X can be constant pattern (except 'null'), Declaration pattern, Recursive pattern.
+            // The PatternSyntax appears e.g. in switch arms and is different from IsPatternSyntax.
+            static bool IsNegativePatternMatch(PatternSyntaxWrapper patternSyntaxWrapper) =>
+                patternSyntaxWrapper.IsNot() && !patternSyntaxWrapper.IsNull();
+        }
+
+        // expectedAffirmative
```

Comment:
In my opinion such a comment does not look good without a context. But I have no idea how to make it better.